### PR TITLE
Jay - Change Date on Intangible Time Entry

### PR DESCRIPTION
**Pull request by Alforoan**
# Description
Added date verification logic such that if the date is in the future, it gives an error.

## Related PRS (if any):
To test this backend PR you need to checkout the [#1865](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865) frontend PR.
…

## Main changes explained:
- Add controller logic so that the user cannot change the date to an invalid date (date in the future).
…

## How to test:
1. check into current branch
2. do `npm install`, `npm run build` and `npm run start` to run this PR locally
3. Clear site data/cache

### CODE REVIEW

**Review by Pandani07 — APPROVED** [review]
Tested this PR and left comments on the frontend PR. Works as intended

**Review by Renanluizssx — COMMENTED** [review]
see my review @Alforoan please:https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865

**Review by swathyjayaseelan — APPROVED** [review]

**Comment by Coops023** [comment]
I have left my comments in the frontend PR.

**Review by XiaohanMeng — COMMENTED** [review]
Hi, I've left my comment in the frontend project. https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865

**Review by Jingyii800 — APPROVED** [review]
Congratulations! This backend PR works well. Please check more details on this comment. https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865#pullrequestreview-1894916238

**Review by skaush21 — APPROVED** [review]
left a comment on PR 1865: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865]

**Review by MALAV1308 — APPROVED** [review]
Hello , I have tested your PR and backend is working as per requirement. Good Job!

**Review by brandta-1 — APPROVED** [review]
# Review

See [front-end review](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865#pullrequestreview-1912817153).

**Review by SRamen1999 — COMMENTED** [review]
left a comment on PR 1865: https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865%5D

**Comment by Weiyao-Li** [comment]
Left comment at Frontend

File: src/controllers/timeEntryController.js [hunk 1]
```diff
@@ -365,10 +369,23 @@ const timeEntrycontroller = function (TimeEntry) {
...
+  }
+
   const postTimeEntry = async function (req, res) {
+    
+    const currentDate = getCurrentDate();
+    const isValidDate = currentDate >= req.body.dateOfWork;
     const isInvalid = !req.body.dateOfWork
       || !moment(req.body.dateOfWork).isValid()
-      || !req.body.timeSpent;
+      || !req.body.timeSpent || !isValidDate;
```
Comment (nathanah) [inline src/controllers/timeEntryController.js:388]: Should verify permission of valid, past date here.

File: src/controllers/timeEntryController.js [hunk 2]
```diff
@@ -189,8 +189,12 @@ const timeEntrycontroller = function (TimeEntry) {
     const isForAuthUser = personId === req.body.requestor.requestorId;
     const isSameDayTimeEntry = moment().tz('America/Los_Angeles').format('YYYY-MM-DD') === newDateOfWork;
     const canEdit = (await hasPermission(req.body.requestor, 'editTimeEntry')) || (isForAuthUser && isSameDayTimeEntry);
+    const canEditTimeEntryToggleTangible = await hasPermission(
+      req.body.requestor,
+      "editTimeEntryToggleTangible"
+    );
 
-    if (!canEdit) {
+    if (!(canEdit || canEditTimeEntryToggleTangible) ) {
```
Comment (nathanah) [inline src/controllers/timeEntryController.js:197]: Needs to be separated to only affect the relevant data.

File: src/controllers/timeEntryController.js [hunk 3]
```diff
@@ -365,10 +369,23 @@ const timeEntrycontroller = function (TimeEntry) {
...
+    const month = String(currentDate.getMonth() + 1).padStart(2, "0");
+    const day = String(currentDate.getDate()).padStart(2, "0");
+    const formattedDate = `${year}-${month}-${day}`;
+
+    return formattedDate;
+  }
+
   const postTimeEntry = async function (req, res) {
+    
+    const currentDate = getCurrentDate();
```
Comment (nathanah) [inline src/controllers/timeEntryController.js:384]:
```suggestion
    const currentDate = moment().tz('America/Los_Angeles').format('YYYY-MM-DD');
```
Use `moment()` as used elsewhere (e.g. line 190)

File: src/controllers/timeEntryController.js [hunk 4]
```diff
@@ -365,10 +369,23 @@ const timeEntrycontroller = function (TimeEntry) {
...
 
+  const getCurrentDate = () => {
+    const currentDate = new Date();
+    const year = currentDate.getFullYear();
+    const month = String(currentDate.getMonth() + 1).padStart(2, "0");
+    const day = String(currentDate.getDate()).padStart(2, "0");
+    const formattedDate = `${year}-${month}-${day}`;
+
+    return formattedDate;
+  }
```
Comment (nathanah) [inline src/controllers/timeEntryController.js:380]:
```suggestion
```
Add this in a utils file if it's needed, but you can just use moment() as used elsewhere in this file.

File: src/controllers/userProfileController.js [hunk 5]
```diff
@@ -276,9 +276,12 @@ const userProfileController = function (UserProfile) {
   const putUserProfile = async function (req, res) {
     const userid = req.params.userId;
     const isRequestorAuthorized = !!(
-      canRequestorUpdateUser(req.body.requestor.requestorId, userid)
-      && ((await hasPermission(req.body.requestor, 'putUserProfile'))
-        || req.body.requestor.requestorId === userid)
+
+      canRequestorUpdateUser(req.body.requestor.requestorId, userid) && (
+        await hasPermission(req.body.requestor, 'putUserProfile')
+        || await hasPermission(req.body.requestor, 'editTimeEntryToggleTangible') || req.body.requestor.requestorId === userid
```
Comment (nathanah) [inline src/controllers/userProfileController.js:282]:
```suggestion
        || req.body.requestor.requestorId === userid
```
Is this needed?

**Review by nathanah — CHANGES_REQUESTED** [review]
The frontend PR is for `editTimelogDate`, but here, you're using `editTimeEntryToggleTangible`.

**Review by hetvi11110 — COMMENTED** [review]
I have tested this PR with PR# 1865 on the front end. Please take a look at my comment there.

**Review by iven-yao — COMMENTED** [review]
I reviewed this pr with corresponding [frontend](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865#pullrequestreview-1947914171)

**Review by Tiantian-C — APPROVED** [review]

**Review by Tiantian-C — COMMENTED** [review]
Hi, I have tested the PR and left the comment at the frontend. It works, thank you!

**Review by kylepham — CHANGES_REQUESTED** [review]
PR worked as intended, but there is a bug that caused by entering a future date submission followed by a past/current date submission. It created 2 entries in the database: one for the future date, and one for the past/current date. Please see my frontend PR's [review](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865#pullrequestreview-1970399835) for reference.

**Comment by MQChen211** [comment]
I tested this PR with the frontend [https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1865#issuecomment-2040315272](url)

**Pull request closed without merging**